src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -285,122 +285,130 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
                 return expression;
 
             BoundExpression result;
-            switch (expression)
+            try
```

Comment:
I went through the helpers called from this method and added an `EnsureSufficientExecutionStack` for all the ones that are specifically about error recovery. For the general ones, I didn't want to cause `InsufficientStackException`s to be thrown in other paths not prepared to handle them.